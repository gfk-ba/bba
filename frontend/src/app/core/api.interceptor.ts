import { Injectable } from '@angular/core';
import { Reset } from '@dcs/ngx-tools';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { catchError, flatMap, map, take } from 'rxjs/operators';
import { State } from '../reducers';
import { tokenSelector } from '../reducers/login/login.selectors';
import {
  HttpEvent,
  HttpHandler,
  HttpInterceptor,
  HttpRequest,
  HttpErrorResponse,
} from '@angular/common/http';

@Injectable()
export class ApiInterceptor implements HttpInterceptor {
  constructor(private store: Store<State>) {}

  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return this.store.pipe(
      take(1),
      map(state => tokenSelector(state)),
      flatMap(token => {
        request = request.clone({
          setHeaders: {
            Authorization: `Bearer ${token.accessToken}`,
          },
        });
        return next.handle(request).pipe(
          catchError((e: HttpErrorResponse) => {
            if (e.status === 401) {
              this.store.dispatch(new Reset());
            }
            throw e;
          })
        );
      })
    );
  }
}
