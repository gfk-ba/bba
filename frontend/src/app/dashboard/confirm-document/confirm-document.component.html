<div>
  <div class="body row" *ngIf="document">
    <div class="col-6">Project:</div>
    <div class="col-6">{{ document.project.name }}</div>
    <div class="col-6">Type:</div>
    <div class="col-6">{{ document.kindLabel }}</div>
    <div class="col-6">Linked to:</div>
    <div class="col-6">{{ document.linksText }}</div>
    <div class="col-6">Hash:</div>
    <div class="col-6">{{ document.hash }}</div>
    <div class="col-6">Description:</div>
    <div class="col-6">{{ document.description }}</div>
    <div class="col-6">Cost:</div>
    <div class="col-6">{{ 1 | currency:'EUR' }}</div>
    <div class="col-6"></div>
  </div>

  <div class="body row">
    <div class="col-12">
      <h5 class="body-headline">Payment options</h5>
    </div>
    <div class="col-6">
      <form [formGroup]="form" (submit)="validateCoupon()">
          <label for="coupon">Pay with Coupon:&nbsp;</label>
          <input type="text" id="coupon" formControlName="coupon" placeholder="Coupon-ID">
        <button type="submit" class="btn btn-default">Pay with Coupon</button>
        <span *ngIf="couponStatus !== 'untested'">{{ couponStatus }}</span>
      </form>
    </div>
    <div class="col-6 text-right">
      <span class="label">Or pay with:</span>
      <div id="paypal-button-container"></div>
    </div>
  </div>
</div>
