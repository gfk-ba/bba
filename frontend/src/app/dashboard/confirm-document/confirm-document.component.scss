@import '../../../styles/common/variables';

.body {
  .col-6:nth-child(odd) {
    font-weight: bold;
  }
  h5 {
    font-weight: bold;
  }
  label, .label {
    display: block;
    font-weight: bold;
    margin: 0.5 * $default-margin 0;
  }
  input {
    display: block;
    margin-bottom: 0.5 * $default-margin;
  }
  #paypal-button-container {
    text-align: right;
  }
}
