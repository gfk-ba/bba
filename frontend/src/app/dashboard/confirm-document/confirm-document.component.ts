import { HttpClient } from '@angular/common/http';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { APP_ENVIRONMENT, IEnvironment } from '@dcs/ngx-tools';
import { of } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { Document } from '../../reducers/document/models/document.class';
import {
  ChangeDetectionStrategy,
  Component,
  Inject,
  Input,
  ChangeDetectorRef,
  OnChanges,
  Output,
  EventEmitter,
} from '@angular/core';

type CouponStatus = 'valid' | 'invalid' | 'untested';

@Component({
  selector: 'dcs-confirm-document',
  templateUrl: './confirm-document.component.html',
  styleUrls: ['./confirm-document.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ConfirmDocumentComponent implements OnChanges {
  @Input()
  public document: Document;

  @Input()
  public visible = false;

  @Output()
  public paymentSuccessful = new EventEmitter<any>();

  public couponStatus: CouponStatus = 'untested';

  public form: FormGroup;

  constructor(
    fb: FormBuilder,
    private http: HttpClient,
    private cd: ChangeDetectorRef,
    @Inject(APP_ENVIRONMENT) private env: IEnvironment
  ) {
    this.form = fb.group({
      coupon: ['', [Validators.required]],
    });
  }

  public ngOnChanges() {
    if (this.visible) {
      this.setupPaypal();
    }
  }

  public validateCoupon(): void {
    const coupon: string = this.form.value.coupon;
    this.http
      .get(`${this.env.apiUrl}/coupons/${coupon}`, { responseType: 'text' })
      .pipe(
        catchError(() => {
          return of('invalid');
        })
      )
      .subscribe((data: string) => {
        this.couponStatus = data === 'OK' ? 'valid' : 'invalid';
        this.cd.markForCheck();

        if (this.couponStatus === 'valid') {
          this.reset();
          this.paymentSuccessful.next({ coupon });
        }
      });
  }

  private reset(): void {
    this.form.reset();
    this.couponStatus = 'untested';
  }

  private setupPaypal() {
    paypal.Button.render(
      {
        // Set your environment

        env: 'sandbox', // sandbox | production
        // Specify the style of the button

        style: { layout: 'vertical', size: 'medium', shape: 'rect', color: 'gold' },
        // horizontal | vertical // medium | large | responsive // pill | rect // gold | blue | silver | black
        // Specify allowed and disallowed funding sources
        //
        // Options:
        // - paypal.FUNDING.CARD
        // - paypal.FUNDING.CREDIT
        // - paypal.FUNDING.ELV

        funding: { allowed: [paypal.FUNDING.CARD], disallowed: [paypal.FUNDING.CREDIT] },
        // PayPal Client IDs - replace with your own
        // Create a PayPal app: https://developer.paypal.com/developer/applications/create

        client: {
          sandbox:
            'AUogdoy0tztATTKpaSDoTpPbl33y3j1ttp84mHy0PAIXh5UeZppm6Ish6gzX7XEK5ywP_ukFFhQ-Bd9O',
          production: '<insert production client id>',
        },
        payment: function(_data: any, actions: any) {
          return actions.payment.create({
            payment: {
              transactions: [
                {
                  amount: { total: '1.00', currency: 'EUR' },
                },
              ],
            },
          });
        },
        onAuthorize: (_data: any, actions: any) => {
          return actions.payment.execute().then((result: any) => {
            this.paymentSuccessful.next({ paypal: result });
          });

          // Example success response
          // const resp = {
          //   id: 'PAY-0WX04483GR9639255LOSL3EI',
          //   intent: 'sale',
          //   state: 'approved',
          //   cart: '6FA78742GD426071A',
          //   create_time: '2018-09-21T09:45:32Z',
          //   payer: {
          //     payment_method: 'paypal',
          //     status: 'VERIFIED',
          //     payer_info: {
          //       email: 'somebody@some-domain.com',
          //       first_name: 'test',
          //       middle_name: 'test',
          //       last_name: 'buyer',
          //       payer_id: '8629FGNH6Y8X6',
          //       country_code: 'DE',
          //       shipping_address: {
          //         recipient_name: 'test buyer',
          //         line1: 'ESpachstr. 1',
          //         city: 'Freiburg',
          //         state: 'Empty',
          //         postal_code: '79111',
          //         country_code: 'DE',
          //       },
          //     },
          //   },
          //   transactions: [
          //     {
          //       amount: { total: '0.01', currency: 'USD', details: {} },
          //       item_list: {},
          //       related_resources: [
          //         {
          //           sale: {
          //             id: '4HW43255PE5672725',
          //             state: 'pending',
          //             payment_mode: 'INSTANT_TRANSFER',
          //             protection_eligibility: 'ELIGIBLE',
          //             parent_payment: 'PAY-0WX04483GR9639255LOSL3EI',
          //             create_time: '2018-09-21T09:45:32Z',
          //             update_time: '2018-09-21T09:45:32Z',
          //             reason_code: 'RECEIVING_PREFERENCE_MANDATES_MANUAL_ACTION',
          //             amount: { total: '0.01', currency: 'USD', details: { subtotal: '0.01' } },
          //           },
          //         },
          //       ],
          //     },
          //   ],
          // };
        },
      },
      '#paypal-button-container'
    );
  }
}
