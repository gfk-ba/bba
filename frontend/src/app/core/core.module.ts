import { registerLocaleData } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import localeDe from '@angular/common/locales/de';
import { InjectionToken, LOCALE_ID, NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { ServiceWorkerModule } from '@angular/service-worker';
import { APP_ENVIRONMENT } from '@dcs/ngx-tools';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { ActionReducerMap, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { ApiInterceptor } from './api.interceptor';
import { environment } from '../../environments/environment';
import { metaReducers, reducers, State } from '../reducers';
import { LoginGuard } from '../reducers/login/login.guard';

export const REDUCER_TOKEN = new InjectionToken<ActionReducerMap<State>>('Registered Reducers');

export function getReducers() {
  return reducers;
}

@NgModule({
  imports: [
    BrowserModule,
    HttpClientModule,
    StoreModule.forRoot(REDUCER_TOKEN, { metaReducers }),
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    StoreRouterConnectingModule.forRoot({
      stateKey: 'router',
    }),
    ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production }),
  ],
  declarations: [],
  providers: [
    { provide: LOCALE_ID, useValue: 'en' },
    { provide: APP_ENVIRONMENT, useValue: environment },
    {
      provide: REDUCER_TOKEN,
      useFactory: getReducers,
    },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: ApiInterceptor,
      multi: true,
    },
    LoginGuard,
  ],
})
export class CoreModule {
  constructor() {
    registerLocaleData(localeDe);
  }
}
