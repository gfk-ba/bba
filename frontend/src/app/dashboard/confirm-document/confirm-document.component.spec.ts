import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { APP_ENVIRONMENT } from '@dcs/ngx-tools';
import { ConfirmDocumentComponent } from './confirm-document.component';

describe('ConfirmDocumentComponent', () => {
  let component: ConfirmDocumentComponent;
  let fixture: ComponentFixture<ConfirmDocumentComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ConfirmDocumentComponent],
      schemas: [NO_ERRORS_SCHEMA],
      imports: [ReactiveFormsModule, HttpClientTestingModule],
      providers: [{ provide: APP_ENVIRONMENT, useValue: { apiUrl: 'http://localhost:3000' } }],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(ConfirmDocumentComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('reset', () => {
    beforeEach(() => {
      component.form.setValue({ coupon: 'any coupon' });
      component.couponStatus = 'valid';
    });

    it('resets data', () => {
      (component as any).reset();

      expect(component.form.value).toMatchSnapshot();
      expect(component.couponStatus).toEqual('untested');
    });
  });

  describe('validateCoupon', () => {
    let httpMock: HttpTestingController;
    let paymentSuccessfulSpy: jest.SpyInstance;
    let resetSpy: jest.SpyInstance;

    beforeEach(() => {
      httpMock = TestBed.get(HttpTestingController);
      paymentSuccessfulSpy = jest.spyOn(component.paymentSuccessful, 'next');
      resetSpy = jest.spyOn(component as any, 'reset');
    });

    afterEach(() => {
      httpMock.verify();
    });

    it('component has an initial couponStatus of untexted', () => {
      expect(component.couponStatus).toEqual('untested');
    });

    describe('for a valid coupon', () => {
      beforeEach(() => {
        component.form.setValue({ coupon: 'test-coupon' });
        component.validateCoupon();

        const req = httpMock.expectOne('http://localhost:3000/coupons/test-coupon');
        req.flush('OK');
      });

      it('emits on paymentSuccessful', () => {
        expect(paymentSuccessfulSpy).toHaveBeenCalledWith({ coupon: 'test-coupon' });
      });

      it('resets component data', () => {
        expect(resetSpy).toHaveBeenCalled();
      });
    });

    describe('for an invalid coupon', () => {
      beforeEach(() => {
        component.form.setValue({ coupon: 'test-coupon' });
        component.validateCoupon();

        const req = httpMock.expectOne('http://localhost:3000/coupons/test-coupon');
        req.flush('ARGH', { status: 422, statusText: 'nope' });
      });

      it('does not emit on paymentSuccessful', () => {
        expect(paymentSuccessfulSpy).not.toHaveBeenCalled();
      });

      it('sets couponStatus to invalid', () => {
        expect(component.couponStatus).toEqual('invalid');
      });
    });
  });
});
