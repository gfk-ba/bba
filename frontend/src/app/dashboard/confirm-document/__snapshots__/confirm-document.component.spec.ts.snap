// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDocumentComponent reset resets data 1`] = `
Object {
  "coupon": null,
}
`;
